axlearn/common/checkpointer_orbax_emergency_replicator.py

```diff
@@ -0,0 +1,323 @@
+            logging.info("The newly created replicator.yaml was not deleted in time.")
+        else:
+            logging.info("The newly created replicator.yaml was deleted, moving forward.")
+            _block_and_process_restore_dir(self._local_dir)
+
+        self._tensor_manager: Optional[oercp.ReplicatorCheckpointManager] = None
+        # See comments of _eval_summaries in `OrbaxCheckpointer`.
+        self._eval_summaries = None
+        self._reached_preemption = False
+
+    def _get_abstract_state(
+        self, state_with_tensors: Nested[Tensor]
+    ) -> Nested[jax.ShapeDtypeStruct]:
+        """Generate the abstract states required by the Orbax replicator checkpointer."""
+        return jax.tree.map(
+            lambda x: jax.ShapeDtypeStruct(x.shape, x.dtype, sharding=x.sharding),
+            state_with_tensors,
+        )
+
+    def _get_tensor_manager(
+        self,
+    ) -> oercp.ReplicatorCheckpointManager:
+        """Creates the replicator checkpoint manager if not exists.
+
+        We defer the creation of this checkpoint manager because it requires the state dict,
+        which is not present during __init__.
+        """
+        if self._tensor_manager is not None:
+            return self._tensor_manager
+
+        # For meaning of these options, refer to
+        # https://github.com/google/orbax/blob/de0b6d0bca643d12840ae73a1f7cfee80af73dcd/checkpoint/orbax/checkpoint/experimental/emergency/replicator_checkpoint_manager.py#L87
+        self._tensor_manager = oercp.ReplicatorCheckpointManager(
+            self._local_dir,
+            options=oercp.ReplicatorCheckpointManagerOptions(
+                save_interval_steps=100,
+                step_name_format=self._name_format,
+            ),
+            global_mesh=thread_resources.env.physical_mesh,
+        )
+        return self._tensor_manager
+
+    def save(
+        self, *, step: int, state: Nested[Tensor], evaler_summaries: Optional[Dict[str, Any]] = None
+    ):
+        """See `BaseCheckpointer.save` for details."""
+        assert self._eval_summaries is None, self._eval_summaries
+        self._eval_summaries = copy.deepcopy(evaler_summaries or {})
+        self._reached_preemption = self._tensor_manager.reached_preemption(step)
+
+        state_with_tensors = jax.tree.map(
+            lambda x: x if isinstance(x, (Tensor, TensorSpec)) else None, state
+        )
+
+        start_t = time.perf_counter()
+        self._get_tensor_manager().save(
+            step=step, args=ocp.args.Composite(state=ocp.args.PyTreeSave(item=state_with_tensors))
+        )
+        time_diff = time.perf_counter() - start_t
+        logging.info("Save time is %fs.", time_diff)
+        self._eval_summaries = None
+        if self._reached_preemption:
+            self.wait_until_finished()
+            raise SystemExit(f"Exiting after saving checkpoint at {step=} due to pre-emption.")
+
+    def _checkpoint_steps_include_local(self) -> list[int]:
```

Comment:
While it does not hurt, realistically we don't need this function.
`.restore` created by Replicator specifies exactly which step to restore.